File: 2_Day/CODE/code.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        i, j = 0, len(s) - 1

        def is_alphanumeric(ch):
            code = ord(ch)
            # 0–9: 48–57, A–Z: 65–90, a–z: 97–122
            return (48 <= code <= 57) or (65 <= code <= 90) or (97 <= code <= 122)

        while i < j:
            left, right = s[i], s[j]

            if not is_alphanumeric(left):
                i += 1
                continue

            if not is_alphanumeric(right):
                j -= 1
                continue

            if left.lower() != right.lower():
                return False

            i += 1
            j -= 1

        return True

File: 2_Day/CODE/code.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function(s) {
    let i = 0;
    let j = s.length - 1;

    function isAlphaNumeric(ch) {
        const code = ch.charCodeAt(0);
        // 0–9: 48–57, A–Z: 65–90, a–z: 97–122
        return (
            (code >= 48 && code <= 57) ||
            (code >= 65 && code <= 90) ||
            (code >= 97 && code <= 122)
        );
    }

    while (i < j) {
        let left = s[i];
        let right = s[j];

        if (!isAlphaNumeric(left)) {
            i++;
            continue;
        }

        if (!isAlphaNumeric(right)) {
            j--;
            continue;
        }

        if (left.toLowerCase() !== right.toLowerCase()) {
            return false;
        }

        i++;
        j--;
    }

    return true;
};
